README.rst
Introduction
============

This package provides a generic view with multiple tabs for plone. It
provides a generic base tab for listing contents in a table, based on
`ftw.table`_.


Features
========

- Generic tabbed view
- Tabs are registered through FTI actions
- Base view for listing tabs
- Listing tabs are filterable
- Perform configurable actions on listed items
- `ftw.table`_'s `Ext JS`_ support works also in listing tables
- Fallback tables
- Drag'n drop multiple file upload functionality (using quickupload plugin)


Installation
============


**Default table implementation**

- Add ``ftw.tabbedview`` to your buildout (or as dependency to a custom egg):

::

    [buildout]
    parts =
        instance
        ...

    [instance]
    ...
    eggs +=
        Plone
        ftw.tabbedview

- Install default profile in portal_setup.


**Ext JS table implementation**

- Add ``ftw.tabbedview`` to your buildout (or as dependency to a custom egg),
  using the ``extjs`` extras require:

::

    [buildout]
    parts =
        instance
        ...

    [instance]
    ...
    eggs +=
        Plone
        ftw.tabbedview[extjs]

- Install extjs profile in portal_setup.


Quickupload plugin implementation
---------------------------------

The quickupload plugin integrates the `collective.quickupload`_ packages in to the tabbedview.

- Add ``ftw.tabbedview`` to your buildout (or as dependency to a custom egg),
  using the ``quickupload`` extras require:

::

    [buildout]
    parts =
        instance
        ...

    [instance]
    ...
    eggs +=
        Plone
        ftw.tabbedview[quickupload]

- Install quickupload profile in portal_setup.

- For activating the quickupload plugin on a context, make sure the context provides the ITabbedviewUploadable Interface.

=====
Usage
=====

We use the package ``example.conference``_ as example for showing how to use ``ftw.tabbedview``.

- Use the ``@@tabbed_view`` on any container.

- Define actions on the content type FTI (Example: ``profiles/default/types/example.conference.program.xml``)::

    <?xml version="1.0"?>
    <object name="example.conference.program" meta_type="Dexterity FTI"
            i18n:domain="example.conference" xmlns:i18n="http://xml.zope.org/namespaces/i18n">

      <property name="default_view">tabbed_view</property>
      <property name="view_methods">
          <element value="tabbed_view"/>
      </property>

      <action title="Sessions" action_id="sessions" category="tabbedview-tabs"
              condition_expr="" url_expr="string:${object_url}?view=sessions"
              visible="True">
          <permission value="View"/>
      </action>

    </object>

- Create the "tab" view (Example: ``browser/tabs.py``)::

    >>> from ftw.tabbedview.browser.listing import CatalogListingView
    >>> from ftw.table import helper
    >>> from example.conference import _
    >>>
    >>> class SessionsTab(CatalogListingView):
    ...     """A tabbed-view tab listing sessions on a program.
    ...     """
    ...
    ...     types = ['example.conference.session']
    ...     sort_on = 'sortable_title'
    ...
    ...     show_selects = False
    ...
    ...     columns = (
    ...         {'column': 'Title',
    ...          'sort_index': 'sortable_title',
    ...          'column_title': _(u'Title'),
    ...          'helper': helper.linked},
    ...
    ...         {'column': 'Track',
    ...          'column_title': _(u"Track")},
    ...         )

- Register the view using ZCML, be sure to name it ``tabbedview_view-${action id}``
  (Example: ``browser/configure.zcml``)::

    <configure
        xmlns="http://namespaces.zope.org/zope"
        xmlns:browser="http://namespaces.zope.org/browser">

        <browser:page
            for="example.conference.program.IProgram"
            name="tabbedview_view-sessions"
            class=".tabs.SessionsTab"
            permission="zope2.View"
            />

    </configure>


Alternative listing sources
===========================

It is possible to use alternative sources for listing tabs. The tables are generated
using `ftw.table`_ and the tab is a ``ftw.table.interfaces.ITableSourceConfig``, which
allows ``ftw.table`` to find an appropriate source. Subclassing ``ITableSourceConfig`` and
registering a custom ``ITableSource`` multi adapter makes it possible to use alternative
data sources such as sqlalchemy or structured python data (local roles for instance).
Take a look at the `ftw.table`_ documentation for more details.


Screenshots
===========

Screenshot of a example tabbed view using the default table implementation:

.. image:: https://github.com/4teamwork/ftw.tabbedview/raw/master/docs/screenshot1.png

Screenshot of the same listing using the ``extjs`` table implementation:

.. image:: https://github.com/4teamwork/ftw.tabbedview/raw/master/docs/screenshot2.png


Caching
=======

``ftw.tabbedview`` provides a ``plone.app.caching`` etag adapter
named ``tabbedview``.
This etag can be used in the caching configuration in order to make the cache
flush when changing the default tab.

In order to enable this, you need to add it to your caching configuration.
Depending on how you want to set up caching in your project, you may want
to add the etag to your rulesets.
You can do that in a ``registry.xml``:

.. code:: xml

    <?xml version="1.0"?>
    <registry>

        <record name="plone.app.caching.weakCaching.plone.content.itemView.etags">
            <value purge="False">
                <element>tabbedview</element>
            </value>
        </record>

        <record name="plone.app.caching.weakCaching.plone.content.folderView.etags">
            <value purge="False">
                <element>tabbedview</element>
            </value>
        </record>

    </registry>


Links
=====

- Github: https://github.com/4teamwork/ftw.tabbedview
- Issues: https://github.com/4teamwork/ftw.tabbedview/issues
- Pypi: http://pypi.python.org/pypi/ftw.tabbedview
- Continuous integration: https://jenkins.4teamwork.ch/search?q=ftw.tabbedview



Licensing
=========

This package is released under GPL Version 2.
Be aware, that when using the package with the ``extjs`` extras, it will
install `Ext JS`_, which has different license policies. See
http://www.sencha.com/products/extjs/license/ for details.


Copyright
=========

This package is copyright by `4teamwork <http://www.4teamwork.ch/>`_.

``ftw.tabbedview`` is licensed under GNU General Public License, version 2.


.. _ftw.table: https://github.com/4teamwork/ftw.table
.. _example.conference: https://github.com/collective/example.conference
.. _Ext JS: http://www.sencha.com/products/extjs/
.. _collective.quickupload: https://github.com/collective/collective.quickupload


docs/HISTORY.txt
Changelog
=========


4.2.2 (unreleased)
------------------

- Nothing changed yet.


4.2.1 (2020-10-23)
------------------

- Fix bug where empty action lists would cause a failure to render the tab. [lgraf]


4.2.0 (2019-11-26)
------------------

- Drop Plone 4.2 support. [jone]
- Fix CSS (quotes) so that is Plone 5 less compatible. [jone]


4.1.3 (2019-08-08)
------------------

- Also show item count for "all" selection in French. [njohner]


4.1.2 (2018-10-24)
------------------

- Better support for touch devices when selecting a tab. [Kevin Bieri]


4.1.1 (2018-04-27)
------------------

- Fix bug which caused unsortable columns to be sortable after storing grid state. [jone]

4.1.0 (2018-02-16)
------------------

- Fix filter for IE when using `clear field` button. [Kevin Bieri]
- Add support for Plone 5.1 [njohner]


4.0.0 (2018-01-24)
------------------

- Use classes instead ids for batching-template to prevent duplicated ids [elioschmutz]

- Load the tabbedview itself async. [Rotonen]


3.8.3 (2017-08-02)
------------------

- Fix: Avoid calculating action menu buttons three times. [buchi]


3.8.2 (2017-06-27)
------------------

- Fix: no longer render the viewname as html on the fallback view. [mathias.leimgruber]


3.8.1 (2017-04-24)
------------------

- Refactor DefaultGridStateStorageKeyGenerator to improve overridable in more
  specific implemnetations.
  [elioschmutz]


3.8.0 (2017-03-22)
------------------

- Append action object name/id to CSS classes in folder action menus. [Rotonen]


3.7.2 (2017-03-13)
------------------

- Fix cachekey / defaulttab generation for anonymous users. [jone]


3.7.1 (2017-03-13)
------------------

- Fix error when calculating etag value on non-tabbedview. [jone]


3.7.0 (2017-03-02)
------------------

- Add plone.app.caching etag value adapter containing the users default tab.
  In order to benefit from that, you need to update your caching configuration.
  See the readme for instructions. [jone]


3.6.2 (2017-03-02)
------------------

- Fix default tab for HTTPS. [mbaechtold]


3.6.1 (2017-03-01)
------------------

- Fix default tab for HTTPS. [jone]


3.6.0 (2016-12-06)
------------------

- Change markup of batching so it can be styled more easily.
  [mbaechtold]

- Template API breakage - renamed slot 'content-title' to 'tabbedview-content-title'
  [Rotonen]

- Minor Chameleon compatibility fixes.
  [Rotonen]


3.5.4 (2016-10-18)
------------------

- Fix tabbedview history.
  [Kevin Bieri]


3.5.3 (2016-09-27)
------------------

- Fix NameError in tabebdview macros. [jone]


3.5.2 (2016-09-26)
------------------

- Chameleon support: make macros access lazy. [jone]

- Check for searchbox value when a tabbedview page gets reloaded.
  [phgross]


3.5.1 (2016-08-23)
------------------

- Extract tabbedview's allowed attributes in to an interface.
  [phgross]

- Move uploadbox above tabbedview menu
  [Kevin Bieri]

- Change Regex in parse_param. Only match decimal characters, since parseFloat may return wrong results when confronted with an int.
  [tschanzt]


3.5.0 (2016-03-07)
------------------

- Drop support for Plone 4.1.x.
  [deiferni]

- Respect permissions defined on tabs during tab-lookup.
  Replaces adapter lookup with traversal to make sure that permissions
  defined on a tab are respected.
  [deiferni]

- Handle when the client gets a response from different source than a tabbed-view.
  By default show a message that the tab could not be loaded successfully. This can
  be customized by registering a different event handler.
  [deiferni]


3.4.3 (2016-02-15)
------------------

- Prevent reload loop when requesting non-existing tab. [jone]


3.4.2 (2015-09-01)
------------------

- Add fallback in tabbedview initalization, when passed index of
  oneBeforeClick event is incorrect.
  [phgross]

- Fix javascript error in IE10 when no default index is defined.
  [Kevin Bieri]


3.4.1 (2015-08-03)
------------------

- Remove whitespace around content title.
  [deiferni]


3.4.0 (2015-05-26)
------------------

- Move tabbedview specific styles from plonetheme.onegov to this package.
  [lknoepfel]

- Fixed initial tab marker (default-tab).
  [phgross]


3.3.13 (2014-08-28)
-------------------

- Add before-tab slot.
  [mathias.leimgruber]


3.3.12 (2014-08-07)
-------------------

- JS: Use base URL to build absolute URL for set_default_tab.
  [lgraf]


3.3.11 (2014-05-28)
-------------------

- Enable D'n'D upload on tabbedview for IE > 9.
  [mathias.leimgruber]


3.3.10 (2014-05-26)
-------------------

- Made TabbedView title customizable by a custom tabbedview template.
  [phgross]

- Replaced old spinner image.
  [Julian Infanger]

- Use bind() instead of on() to restore jQuery < 1.7 compatibility.
  [lgraf]


3.3.9 (2014-03-28)
------------------

- Fix bug sometimes the default-tab setting did not work.
  [jone]

- Fixed not closed script tags.
  [lknoepfel]

- Load extjs related JS resurces only on tabbed.pt, thus we save
  700KB of JS on not tabbed views.
  [mathias.leimgruber]


3.3.8 (2013-11-28)
------------------

- Replaced unicode char for tabbedview menu header.
  Use icon instead, because the unicode char didn't work on IE.
  [Julian Infanger]


3.3.7 (2013-08-26)
------------------

- Move table_options initalization from __call__ to __init__ in listing.py.
  Otherwise the update method does not work without calling the view first.
  [mathias.leimgruber]

- Fix deprecated quickupload interface import.
  [jone]


3.3.6 (2013-07-18)
------------------

- Filter field: add placeholder.
  [jone]

- Fixed markup in tabbedview buttons and added a class to identify buttons.
  [Julian Infanger]

- Mark tabbedview header with disabledSearchBox class if the searchbox is disabled.
  [Julian Infanger]


3.3.5 (2013-04-17)
------------------

- Use removeAttr (jQuery 1.7.x).
  [mathias.leimgruber]

- Fix reset table configuration action.
  [mathias.leimgruber]

- Use Batch.fromPagenumber.
  [mathias.leimgruber]



ftw/tabbedview/utils.py
from datetime import datetime
from datetime import timedelta
import re


def get_filter_values(columns, request, filter_state):
    """Parses the request, the personal filter_state (gridstate['filter'])
    and the default filter values and returns a dict of the current filters
    to be used.

    Example return value:
    >>> {'getResponsibleManager': {'type': 'list',
    ...                            'value': ['foo', 'bar']},
    ...  'modified': {'type': 'date',
    ...               'value': {
    ...                   'gt': '2012-12-11',
    ...                   'lt': '2012-12-31'}}}
    """

    state_filters = get_filters_from_state(columns, filter_state)
    if state_filters is not None:
        return state_filters

    request_filters = get_filters_from_request(request)
    if request_filters is not None:
        return request_filters

    values = {}
    types_and_defaults = get_column_filter_types_and_defaults(columns)

    for name in types_and_defaults.keys():
        type_, default = types_and_defaults[name]
        if default is None:
            continue

        values[name] = {'type': type_,
                        'value': default}

    return values


def get_filters_from_request(request):
    """This helper function helps extracting the currently
    active filters from the request. Since the ExtJS filter
    uses request keys which are not automatically parseable
    by Zope, we need to apply some magic.

    Example return value:
    >>> {'getResponsibleManager': {'type': 'list',
    ...                            'value': ['foo', 'bar']}}

    """

    if not request.form:
        return None

    data = {}

    for key, value in request.form.items():
        # keys are in form "filter[0][data][type]"

        if not key.startswith('filter['):
            continue

        key = key.replace('[data]', '')
        match = re.match(r'filter\[(\d)\]\[(.*?)\]', key)
        if not match:
            continue

        index = match.groups()[0]
        name = match.groups()[1]

        if index not in data:
            data[index] = {}
        data[index][name] = value


    filters = {}

    def convert_date(datestr):
        # convert from %m/%d/%Y to %Y-%m-%d
        return re.sub(r'^(\d{2})/(\d{2})/(\d{4})$',
                      r'\g<3>-\g<1>-\g<2>',
                      datestr)

    for item in data.values():
        field = item.get('field')
        type_ = item.get('type')

        if field not in filters:
            filters[field] = {
                'type': type_}

        if type_ == 'date':
            if 'value' not in filters[field]:
                filters[field]['value'] = {}

            comparison = item.get('comparison')
            filters[field]['value'][comparison] = convert_date(item['value'])

        elif type_ == 'list' and isinstance(item.get('value'), list):
            new_value = []

            for val in item['value']:
                if isinstance(val, unicode):
                    val = val.encode('utf-8')
                new_value.append(val)

            filters[field]['value'] = new_value

        else:
            filters[field]['value'] = item.get('value')

    return filters or None


def get_filters_from_state(columns, filter_state):
    if filter_state is None:
        return None

    types_and_defaults = get_column_filter_types_and_defaults(columns)

    result = {}

    def convert_date(datestr):
        # XXX: the date in the persistent filter state is the day before at
        # 23:00, because of an timezone offset. We fix it manually here.
        # convert from: '2012-12-17T23:00:00.000Z'
        # to: '2012-12-18'

        # remove ".000Z" at the end
        datestr = datestr.split('.')[0]

        # parse and fix offset
        date = datetime.strptime(datestr, '%Y-%m-%dT%H:%M:%S')
        date += timedelta(hours=1)

        # format
        return date.strftime('%Y-%m-%d')

    for key, value in filter_state.items():
        if key not in types_and_defaults:
            continue

        type_, _default = types_and_defaults[key]

        if type_ == 'date':
            date_value = {}

            if 'before' in value:
                date_value['lt'] = convert_date(value['before'])

            if 'after' in value:
                date_value['gt'] = convert_date(value['after'])

            if 'on' in value:
                date_value['eq'] = convert_date(value['on'])

            value = date_value

        elif type_ == 'list':
            if isinstance(value, list):
                new_value = []

                for val in value:
                    if isinstance(val, unicode):
                        val = val.encode('utf-8')
                    new_value.append(val)

                value = new_value

        result[key] = {'type': type_,
                       'value': value}

    return result


def get_column_filter_types_and_defaults(columns):
    filters = {}

    for col in columns:
        if not isinstance(col, dict):
            continue

        col_id = col.get('sort_index', col.get('column'))
        if not col_id:
            continue

        filter_ = col.get('filter')
        if not filter_:
            continue

        definition = filter_.get_filter_definition(col, [])
        if not definition or not definition.get('type'):
            continue

        default = filter_.get_default_value(col)
        filters[col_id] = (definition['type'], default)

    return filters
